### PR TITLE
Support for BigQuery `struct`,  `array` and `bytes` , `int64`, `float64` datatypes

### LINKED ISSUE #901: Support field names in BigQuery STRUCT
I want to parse BigQuery queries that **assign field names** in their STRUCTs.  
Example:
```SQL
SELECT
    STRUCT(
        CONCAT(user.first_name, ' ', user.last_name) AS full_name,
        user.email AS email,
    ) AS viewer
FROM users
```

This corresponds to the typeless struct syntax mentioned in the [BigQuery documentation](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#typeless_struct_syntax), and which is as follows:  
 `STRUCT( expr1 [AS field_name] [, ... ])`

Note that without the aliases, the query is parsed successfully:
```SQL
SELECT
    STRUCT(
        CONCAT(user.first_name, ' ', user.last_name),
        user.email
    ) AS viewer
FROM users
```

It seems there is already [an ongoing PR](https://github.com/sqlparser-rs/sqlparser-rs/pull/817) related to BigQuery STRUCTs, but the description does not mention support for field names/aliases.

### PR DESCRIPTION
**Pull request by iffyio**
This builds on top of #817

- `STRUCT` literal support via `STRUCT` keyword https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#constructing_a_struct
- `STRUCT` and `ARRAY` type declarations https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#declaring_an_array_type https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#declaring_a_struct_type

It works around the issue of not being able to parse nested types like `ARRAY<ARRAY<INT>>` due to the right angle bracket ambiguity where the tokenizer chooses the right-shift operator (this affects other dialects like Hive that have similar syntax). When parsing such types, we accept a closing `>` or `>>` and track which variant is in use in order to preserve correctness.

Fixes #901
Closes #817

### CODE REVIEW

**Comment by alamb** [comment]
Note there appears to be a similar PR: https://github.com/sqlparser-rs/sqlparser-rs/pull/966

**Review by alamb — APPROVED** [review]
Thank you very much for this contribution @iffyio 👏   -- it is very impressive both well documented and well tested. I had a few minor comments, but nothing that would prevent merging.

Please let me know if you have time to address the comments this week (I plan to make a release later this week)

**Comment by iffyio** [comment]
Thanks for the review @alamb! Sure I'll fetch the latest changes and look to address the comments later today :+1: 

> Note there appears to be a similar PR: https://github.com/sqlparser-rs/sqlparser-rs/pull/966

Ah indeed I missed that one, thanks for the heads up!

File: src/parser/mod.rs [hunk 1]
```diff
@@ -816,6 +836,10 @@ impl<'a> Parser<'a> {
                 Keyword::MATCH if dialect_of!(self is MySqlDialect | GenericDialect) => {
                     self.parse_match_against()
                 }
+                Keyword::STRUCT if dialect_of!(self is BigQueryDialect) => {
```
[the remark this answers is not in the document]
Comment (iffyio) [reply src/parser/mod.rs]: Done!

File: src/parser/mod.rs [hunk 2]
```diff
@@ -1728,6 +1752,165 @@ impl<'a> Parser<'a> {
...
+        Ok(Expr::Struct { values, fields })
+    }
+
+    /// Parse an expression value for a bigquery struct [1]
+    /// Syntax
+    /// ```sql
+    /// expr [AS name]
+    /// ```
+    /// [1]: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#constructing_a_struct
+    pub fn parse_struct_field_expr(&mut self, typed_syntax: bool) -> Result<Expr, ParserError> {
```
[the remark this answers is not in the document]
Comment (iffyio) [reply src/parser/mod.rs]: Fixed!

File: src/parser/mod.rs [hunk 3]
```diff
@@ -1728,6 +1752,165 @@ impl<'a> Parser<'a> {
...
+
+        self.expect_token(&Token::LParen)?;
+        let values = self
+            .parse_comma_separated(|parser| parser.parse_struct_field_expr(!fields.is_empty()))?;
+        self.expect_token(&Token::RParen)?;
+
+        Ok(Expr::Struct { values, fields })
+    }
+
+    /// Parse an expression value for a bigquery struct [1]
```
[the remark this answers is not in the document]
Comment (iffyio) [reply src/parser/mod.rs:1848]: Updated the description to cover the parameter

**Comment by alamb** [comment]
Thanks again @iffyio

**Pull request merged**